### PR TITLE
Add new API-reference page, automated generation of `.rst` files, enable `sphinx-gallery`. Fix #1808

### LINKED ISSUE #1808: Reformat API reference page

### PR DESCRIPTION
**Pull request by anotherbugmaster**
Fix #1808 

Reformatted API reference, but there's a little bit ugly workaround there: reference (like `[1]_`) doesn't work correctly if there's another reference with the same number in the upper section. I explored sklearn repo, it seems like they're aware of it and trying not to do so or making the same workaround as me.

[Autosummary sphinx module](http://www.sphinx-doc.org/en/1.5.1/ext/autosummary.html)

[Templating](http://www.sphinx-doc.org/en/1.5.1/templating.html#templating)

### CODE REVIEW

File: gensim/summarization/summarizer.py [hunk 1]
```diff
@@ -322,15 +322,16 @@ def _build_hasheable_corpus(corpus):
 
 
 def summarize_corpus(corpus, ratio=0.2):
-    """Get a list of the most important documents of a corpus using a variation of the TextRank algorithm [1]_.
-     Used as helper for summarize :func:`~gensim.summarization.summarizer.summarizer`
+    """Get a list of the most important documents of a corpus using a
+    variation of the TextRank algorithm [1].
```
Comment (menshikh-iv) [inline gensim/summarization/summarizer.py]: probably, you can use text identifiers for it, for example `[some-paper]`

**Comment by menshikh-iv** [comment]
We'll continue work on it together with @anotherbugmaster in Jan

File: docs/src/Makefile [hunk 2]
```diff
@@ -32,6 +32,7 @@ help:
 
 clean:
 	-rm -rf $(BUILDDIR)/*
+	-rm -rf generated/*
```
[the remark this answers is not in the document]
Comment (anotherbugmaster) [reply docs/src/Makefile]: In no way!

`generated` is the folder which contains new auto-generated `.rst` files. It ought to be purged every time, otherwise *autosummary* will not update reference page.

File: docs/src/_templates/class.rst [hunk 3]
```diff
@@ -0,0 +1,15 @@
+:mod:`{{module}}`.{{objname}}
```
[the remark this answers is not in the document]
Comment (anotherbugmaster) [reply docs/src/_templates/class.rst]: Done

File: gensim/summarization/summarizer.py (hunk 1 above)
Comment (anotherbugmaster) [reply gensim/summarization/summarizer.py]: AFAIK this bug is connected to trailing underscore in `[reference_name]_`. I'll try to find a better workaround.

Comment (menshikh-iv) [reply gensim/summarization/summarizer.py]: @anotherbugmaster add links to comment for this PR about this bug, I'll try to find workaround too

**Comment by anotherbugmaster** [comment]
As for bug message:

`<autosummary>:1:Unknown target name: "1"`

Here's what I found:

https://groups.google.com/forum/#!topic/sphinx-users/JVNPh7AIKi8
https://github.com/Abjad/abjad/issues/332

File: gensim/summarization/summarizer.py (hunk 1 above)
Comment (anotherbugmaster) [reply gensim/summarization/summarizer.py]: Done!

**Comment by markroxor** [comment]
@anotherbugmaster Just a generic gentle suggestion - Can you please change the title to something more informative?  😄

File: docs/_static/js/jquery-migrate-1.1.1.min.js [hunk 4]
```diff
@@ -1,3 +1,3 @@
-/*! jQuery Migrate v1.1.1 | (c) 2005, 2013 jQuery Foundation, Inc. and other contributors | jquery.org/license */
+/*! jQuery Migrate v1.1.1 | (c) 2005, 2013 jQuery Foundation, Inc. and other contributors | jquery.org/license */
```
Comment (menshikh-iv) [inline docs/_static/js/jquery-migrate-1.1.1.min.js]: why diff?

File: docs/_templates/class.rst [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+{{ underline }}==============
+
+.. currentmodule:: {{ module }}
+
+.. autoclass:: {{ objname }}
+    :members:
+    :special-members:
+    :exclude-members: __init__
+
+.. raw:: html
```
Comment (menshikh-iv) [inline docs/_templates/class.rst]: What's do this command?

File: docs/_templates/class.rst [hunk 6]
```diff
@@ -0,0 +1,13 @@
+:mod:`{{module}}`.{{objname}}
+{{ underline }}==============
+
+.. currentmodule:: {{ module }}
+
+.. autoclass:: {{ objname }}
+    :members:
+    :special-members:
+    :exclude-members: __init__
```
Comment (menshikh-iv) [inline docs/_templates/class.rst]: Why exclude `__init__`? Some duplication problems?

File: docs/_templates/class.rst [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+.. raw:: html
+
+    <div class="clearer"></div>
```
Comment (menshikh-iv) [inline docs/_templates/class.rst]: why did this need? looks strange, probably this line should be in template of site, not `rst` template

File: docs/apiref_nextgen.rst [hunk 8]
```diff
@@ -0,0 +1,479 @@
+=============
+API Reference
```
Comment (menshikh-iv) [inline docs/apiref_nextgen.rst]: can you add it to site menu after current apiref?

File: docs/src/gensim_theme/static/doctools.js [hunk 9]
```diff
@@ -1,247 +0,0 @@
-/*
```
Comment (menshikh-iv) [inline docs/src/gensim_theme/static/doctools.js]: suspicious, why this removed (same question for all `*.js` files)

File: docs/user_guide.rst [hunk 10]
```diff
@@ -0,0 +1,11 @@
+.. title:: User guide: contents
```
Comment (menshikh-iv) [inline docs/user_guide.rst]: Need to make new menu tab on site for this too (as for new API)

File: examples/README.txt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+.. _general_examples:
```
Comment (menshikh-iv) [inline examples/README.txt]: what is it? you create new folder in root with this one, probably incorrect location

File: gensim/models/fasttext.py [hunk 12]
```diff
@@ -161,7 +161,7 @@ def train_batch_sg(model, sentences, alpha, work=None, neu1=None):
 
 class FastText(BaseWordEmbeddingsModel):
     """Class for training, using and evaluating word representations learned using method
-    described in [1]_ aka Fasttext.
+    described in [1] aka Fasttext.
```
Comment (menshikh-iv) [inline gensim/models/fasttext.py]: about `[1]_` bug, let's switch to `some text <http://asdfasd.com>`_ format (this is better than the link that doesn't work), here and everywhere.

File: gensim/models/__init__.py [hunk 13]
```diff
@@ -1,6 +1,5 @@
 """
-This package contains algorithms for extracting document representations from their raw
-bag-of-word counts.
+Extraction of documents representations from their raw bag-of-word counts.
```
Comment (menshikh-iv) [inline gensim/models/__init__.py]: Maybe use `"""sometext."""` everywhere in headers `if len(line) <= 120` ?

File: gensim/models/translation_matrix.py [hunk 14]
```diff
@@ -266,8 +266,9 @@ def translate(self, source_words, topn=5, gc=0, sample_num=None, source_lang_vec
         topn : int, optional
             Number of words than will be returned as translation for each `source_words`
         gc : int, optional
-            Define translation algorithm, if `gc == 0` - use standard NN retrieval,
-            otherwise, use globally corrected neighbour retrieval method (as described in [1]_).
+            Define the translation algorithm, if `gc == 0` - use standard NN
+            retrieval,
```
Comment (menshikh-iv) [inline gensim/models/translation_matrix.py]: ??

File: gensim/topic_coherence/__init__.py [hunk 15]
```diff
@@ -1,4 +1,3 @@
 """
-This package contains implementation of the individual components of
-the topic coherence pipeline.
+Individual components of the topic coherence pipeline.
```
Comment (menshikh-iv) [inline gensim/topic_coherence/__init__.py]: question about `__init__.py`: this type of files will be suggested in reference? If yes - how this will looks like?

File: setup.py [hunk 16]
```diff
@@ -309,7 +309,12 @@ def finalize_options(self):
         'distributed': distributed_env,
         'test-win': win_testenv,
         'test': linux_testenv,
-        'docs': linux_testenv + distributed_env + ['sphinx', 'sphinxcontrib-napoleon', 'plotly', 'pattern', 'sphinxcontrib.programoutput'],
+        'docs': linux_testenv + distributed_env + ['sphinx',
```
Comment (menshikh-iv) [inline setup.py]: no vertical indents please (only hanging)

File: gensim/models/fasttext.py (hunk 12 above)
Comment (anotherbugmaster) [reply gensim/models/fasttext.py]: Ok!

File: docs/_static/js/jquery-migrate-1.1.1.min.js (hunk 4 above)
Comment (anotherbugmaster) [reply docs/_static/js/jquery-migrate-1.1.1.min.js]: CRLF probably. I'll look into it.

File: docs/Makefile [hunk 17]
```diff
@@ -6,6 +6,7 @@
 
 # You can set these variables from the command line.
 SPHINXOPTS    = -W
+# SPHINXOPTS =
```
[the remark this answers is not in the document]
Comment (anotherbugmaster) [reply docs/Makefile]: Ok

File: docs/_templates/class.rst (hunk 7 above)
Comment (anotherbugmaster) [reply docs/_templates/class.rst]: Well, I copypasted these templates from sklearn. :)

Comment (anotherbugmaster) [reply docs/_templates/class.rst]: It's working, isn't it? We'll remove redundant parts later.

File: examples/README.txt (hunk 11 above)
Comment (anotherbugmaster) [reply examples/README.txt]: Nope. *examples* directory is the thing for the sphinx-gallery. This line is just a reference name for this file.

File: docs/user_guide.rst (hunk 10 above)
Comment (anotherbugmaster) [reply docs/user_guide.rst]: Will do

File: docs/src/gensim_theme/static/doctools.js (hunk 9 above)
Comment (anotherbugmaster) [reply docs/src/gensim_theme/static/doctools.js]: Because I moved the whole thing to the `../`? :)

File: setup.py (hunk 16 above)
Comment (anotherbugmaster) [reply setup.py]: Ok

File: gensim/topic_coherence/__init__.py (hunk 15 above)
Comment (anotherbugmaster) [reply gensim/topic_coherence/__init__.py]: It will be printed out in the apiref_nextgen.rst

File: gensim/models/__init__.py (hunk 13 above)
Comment (anotherbugmaster) [reply gensim/models/__init__.py]: Ok

File: .gitignore [hunk 18]
```diff
@@ -70,7 +72,9 @@ data
 *.model
 *~
 *_out.txt
-*.html
+docs/*.html
+docs/modules/*.html
 *.inv
 *.js
```
[the remark this answers is not in the document]
Comment (anotherbugmaster) [reply .gitignore]: Nope, it's intended.

File: docs/_templates/class.rst (hunk 7 above)
Comment (menshikh-iv) [reply docs/_templates/class.rst]: If this class isn't defined - this doesn't work. Try to disable & build & compare please

File: examples/README.txt (hunk 11 above)
Comment (menshikh-iv) [reply examples/README.txt]: why this in the root of the repository?  Better to move it do `docs/`

File: docs/src/gensim_theme/static/doctools.js (hunk 9 above)
Comment (menshikh-iv) [reply docs/src/gensim_theme/static/doctools.js]: this **should** be moved (but this removed), check your repository please

File: examples/README.txt (hunk 11 above)
Comment (anotherbugmaster) [reply examples/README.txt]: Well, I can do that. [Scikit-learn](https://github.com/scikit-learn/scikit-learn) and sphinx-gallery themselves place it in the root dir though.

File: docs/_templates/class.rst (hunk 7 above)
Comment (anotherbugmaster) [reply docs/_templates/class.rst]: Ok, will do

**Comment by anotherbugmaster** [comment]
Need to move docs/* back to docs/src/*

File: setup.py (hunk 16 above)
Comment (anotherbugmaster) [reply setup.py]: Done

File: gensim/models/translation_matrix.py (hunk 14 above)
Comment (anotherbugmaster) [reply gensim/models/translation_matrix.py]: Done

File: docs/_templates/class.rst (hunk 7 above)
Comment (anotherbugmaster) [reply docs/_templates/class.rst]: https://forum.dokuwiki.org/thread/10120

File: docs/_templates/class.rst (hunk 5 above)
Comment (anotherbugmaster) [reply docs/_templates/class.rst]: Includes raw html into template. :)

File: docs/Makefile (hunk 17 above)
[the remark this answers is not in the document]
Comment (anotherbugmaster) [reply docs/Makefile]: Done

File: docs/src/Makefile [hunk 19]
```diff
@@ -32,15 +32,18 @@ help:
...
+	-rm -rf modules/generated/
+	-rm -rf auto_examples/*
 
 html:
 	$(SPHINXBUILD) -b html $(ALLSPHINXOPTS) $(BUILDDIR)/html
 	rm -r $(BUILDDIR)/html/_sources
-	cp -r $(BUILDDIR)/html/* ../
+	cp -r $(BUILDDIR)/html/* .
```
Comment (menshikh-iv) [inline docs/src/Makefile]: Why this change?

File: docs/src/_static/js/jquery-migrate-1.1.1.min.js [hunk 20]
```diff
@@ -1,3 +1,3 @@
-/*! jQuery Migrate v1.1.1 | (c) 2005, 2013 jQuery Foundation, Inc. and other contributors | jquery.org/license */
+/*! jQuery Migrate v1.1.1 | (c) 2005, 2013 jQuery Foundation, Inc. and other contributors | jquery.org/license */
```
Comment (menshikh-iv) [inline docs/src/_static/js/jquery-migrate-1.1.1.min.js]: can you simply checkout this file, please?

File: gensim/summarization/summarizer.py [hunk 21]
```diff
@@ -344,6 +345,12 @@ def summarize_corpus(corpus, ratio=0.2):
     list of str
         Most important documents of given `corpus` sorted by the document score, highest first.
 
+    References
+    ----------
+    .. [1] Federico Barrios, Federico L´opez, Luis Argerich, Rosita Wachenchauzer (2016).
```
Comment (menshikh-iv) [inline gensim/summarization/summarizer.py]: `[]`

File: docs/src/_static/js/jquery-migrate-1.1.1.min.js (hunk 20 above)
Comment (anotherbugmaster) [reply docs/src/_static/js/jquery-migrate-1.1.1.min.js]: :D

Comment (anotherbugmaster) [reply docs/src/_static/js/jquery-migrate-1.1.1.min.js]: Sure

File: docs/src/Makefile (hunk 19 above)
Comment (anotherbugmaster) [reply docs/src/Makefile]: My mistake, I don't like the copying to the upper level though. Why are we doing it in the first place?

File: gensim/models/tfidfmodel.py [hunk 22]
```diff
@@ -301,10 +301,12 @@ def __init__(self, corpus=None, id2word=None, dictionary=None, wlocal=utils.iden
             Only when `pivot` is not None will pivoted document length normalization be applied.
             Otherwise, regular TfIdf is used.
         slope : float, optional
-            Parameter required by pivoted document length normalization which determines the slope to which
-            the `old normalization` can be tilted. This parameter only works when pivot is defined.
+            It is the parameter required by pivoted document length normalization which determines the slope to which
```
Comment (piskvorky) [inline gensim/models/tfidfmodel.py]: This conflict is incorrectly resolved. Please continue with changes in #2102, which should be up-to-date with develop.

Comment (menshikh-iv) [reply gensim/models/tfidfmodel.py]: @piskvorky author (@anotherbugmaster) can't commit to your branch (lack of permissions)

Comment (piskvorky) [reply gensim/models/tfidfmodel.py]: Why not add @anotherbugmaster as a committer / collaborator?

Comment (menshikh-iv) [reply gensim/models/tfidfmodel.py]: @piskvorky done
@anotherbugmaster please fix #2102 branch (you have permissions for clone current repository (not your fork), commit & push directly to #2012 branch), sorry for bothering on vacation.

**Comment by menshikh-iv** [comment]
Continue in #2102
